subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildInclusionCoordinator.java

```diff
@@ -82,14 +82,16 @@ public void registerSubstitutionsProvidedBy(BuildState build) {
     }
 
     public void prepareForPluginResolution(IncludedBuildState build) {
-        if (!registering.add(build)) {
-            return;
-        }
-        try {
-            build.ensureProjectsConfigured();
-            makeSubstitutionsAvailableFor(build, new HashSet<>());
-        } finally {
-            registering.remove(build);
+        synchronized (registering) {
```

Comment:
🤔  This synchronization looks suspicious on its own. If it is indeed required, I would make the full service thread-safe by synchronizing on a separate lock object. As it stands there are currently other methods that touch `registering` as well, and they are not synchronized

Reply:
Despite other methods are touching `registering` as well, only currently synchronized access can be triggered from  configuration. Other invocations are happening like at settings level

Reply:
I think "partial" synchronization is harder to reason about, because it requires to know/remember that only specific methods can be called concurrently. This can also change in the future, and the bugs it can introduce can be hard to pin point.

There is no performance penalty for adding synchronization if the other methods are accessed from a single thread.

I am interested in what perspective @bamboo has on this

Reply:
I agree with the notion that partial synchronization is harder to reason about. I'm learning towards synchronizing all code paths that touch `registering`.